File: ones-gfx-sdk-go/ones_gfx/transport.go
package ones_gfx

import (
	"bytes"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"time"
)

// Transport wraps HTTP operations and knows about ONES API conventions
// (envelope unwrapping, Prefer header, error mapping). This is exported so
// the resources package can use it.
type Transport struct {
	baseURL    string
	auth       AuthProvider
	httpClient *http.Client
	timeout    time.Duration
}

// NewTransport constructs a Transport with the given configuration.
func NewTransport(baseURL string, auth AuthProvider, cfg clientConfig) *Transport {
	transport := &http.Transport{
		TLSClientConfig: cfg.tlsConfig,
	}
	if !cfg.verifyTLS {
		transport.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
	}

	httpClient := &http.Client{
		Timeout:   cfg.timeout,
		Transport: transport,
	}

	return &Transport{
		baseURL:    baseURL,
		auth:       auth,
		httpClient: httpClient,
		timeout:    cfg.timeout,
	}
}

// NewTransportWithOptions constructs a Transport using client options.
func NewTransportWithOptions(baseURL string, auth AuthProvider, opts ...ClientOption) *Transport {
	cfg := defaultClientConfig()
	for _, opt := range opts {
		opt(&cfg)
	}
	return NewTransport(baseURL, auth, cfg)
}

// Get performs a GET request and returns the unwrapped data.
func (t *Transport) Get(path string, timeout *time.Duration) (interface{}, error) {
	return t.request("GET", path, nil, OperationModeSynchronous, timeout)
}

// Post performs a POST request with optional mode and timeout override.
func (t *Transport) Post(path string, body interface{}, mode OperationMode, timeout *time.Duration) (interface{}, error) {
	return t.request("POST", path, body, mode, timeout)
}

// Patch performs a PATCH request with optional mode and timeout override.
func (t *Transport) Patch(path string, body interface{}, mode OperationMode, timeout *time.Duration) (interface{}, error) {
	return t.request("PATCH", path, body, mode, timeout)
}

// Delete performs a DELETE request with optional mode and timeout override.
func (t *Transport) Delete(path string, body interface{}, mode OperationMode, timeout *time.Duration) (interface{}, error) {
	return t.request("DELETE", path, body, mode, timeout)
}

// Close releases HTTP resources.
func (t *Transport) Close() error {
	t.httpClient.CloseIdleConnections()
	return t.auth.Close()
}

// request is the internal method that handles auth, retry-on-401, and error mapping.
func (t *Transport) request(method, path string, body interface{}, mode OperationMode, timeout *time.Duration) (interface{}, error) {
	// Build URL
	fullURL, err := t.buildURL(path)
	if err != nil {
		return nil, err
	}

	// Proactive refresh if token is near expiry
	if t.auth.NeedsProactiveRefresh() {
		if err := t.auth.Refresh(); err != nil {
			return nil, err
		}
	}

	// Send request
	resp, err := t.send(method, fullURL, body, mode, timeout)
	if err != nil {
		return nil, err
	}

	// Reactive refresh: if we got a 401, refresh and retry once
	if resp.StatusCode == http.StatusUnauthorized {
		if err := t.auth.Refresh(); err != nil {
			return nil, err
		}
		resp, err = t.send(method, fullURL, body, mode, timeout)
		if err != nil {
			return nil, err
		}
	}

	return t.handleResponse(resp)
}

// send builds headers and sends the HTTP request.
func (t *Transport) send(method, fullURL string, body interface{}, mode OperationMode, timeout *time.Duration) (*http.Response, error) {
	var reqBody io.Reader
	if body != nil {
		bodyBytes, err := json.Marshal(body)
		if err != nil {
			return nil, &TransportError{Message: "failed to marshal request body", Cause: err}
		}
		reqBody = bytes.NewReader(bodyBytes)
	}

	req, err := http.NewRequest(method, fullURL, reqBody)
	if err != nil {
		return nil, &TransportError{Message: "failed to build request", Cause: err}
	}

	// Set headers
	headers := map[string]string{
		"Content-Type": "application/json",
		"Accept":       "application/json",
	}
	t.auth.Apply(headers)

	// Add Prefer header for sync/async mode
	if mode == OperationModeSynchronous {
		headers["Prefer"] = "respond-sync"
	} else if mode == OperationModeAsyncPoll || mode == OperationModeAsyncWebhook {
		headers["Prefer"] = "respond-async"
	}

	for k, v := range headers {
		req.Header.Set(k, v)
	}

	// Apply timeout override if provided
	client := t.httpClient
	if timeout != nil {
		client = &http.Client{
			Timeout:   *timeout,
			Transport: t.httpClient.Transport,
		}
	}

	resp, err := client.Do(req)
	if err != nil {
		return nil, &TransportError{Message: fmt.Sprintf("%s %s failed", method, fullURL), Cause: err}
	}

	return resp, nil
}

// handleResponse parses the response and maps errors.
func (t *Transport) handleResponse(resp *http.Response) (interface{}, error) {
	defer resp.Body.Close()

	bodyBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, &TransportError{Message: "failed to read response body", Cause: err}
	}

	// Parse body as JSON
	var body interface{}
	if len(bodyBytes) > 0 {
		if err := json.Unmarshal(bodyBytes, &body); err != nil {
			// Not JSON - treat as plain text
			body = string(bodyBytes)
		}
	}

	// Handle success
	if resp.StatusCode >= 200 && resp.StatusCode < 300 {
		return unwrapEnvelope(body), nil
	}

	// Handle errors
	message := extractErrorMessage(body)
	if message == "" {
		message = fmt.Sprintf("HTTP %d", resp.StatusCode)
	}

	// Map status code to typed error
	return nil, classifyError(resp.StatusCode, message, body)
}

// buildURL joins baseURL and path.
func (t *Transport) buildURL(path string) (string, error) {
	base, err := url.Parse(t.baseURL)
	if err != nil {
		return "", &TransportError{Message: "invalid base URL", Cause: err}
	}
	rel, err := url.Parse(path)
	if err != nil {
		return "", &TransportError{Message: "invalid path", Cause: err}
	}
	return base.ResolveReference(rel).String(), nil
}

// unwrapEnvelope handles ONES response envelopes.
// Success responses: {"status": "success", "message": "...", "data": ...}
// The data field may be a JSON-encoded string or an object.
func unwrapEnvelope(body interface{}) interface{} {
	obj, ok := body.(map[string]interface{})
	if !ok {
		return body
	}

	// Detect envelope by presence of "status" and "data"/"message"
	status, hasStatus := obj["status"]
	if !hasStatus {
		return body // Not an envelope
	}

	if status != "success" {
		return body
	}

	// Extract data field
	data, hasData := obj["data"]
	if hasData {
		// data may be JSON-encoded string
		if str, ok := data.(string); ok && str != "" {
			var parsed interface{}
			if err := json.Unmarshal([]byte(str), &parsed); err == nil {
				return parsed
			}
			return str
		}
		return data
	}

	// If no data field, return message
	if msg, hasMsg := obj["message"]; hasMsg {
		return msg
	}

	return body
}

// extractErrorMessage pulls a useful error message from the response.
func extractErrorMessage(body interface{}) string {
	if str, ok := body.(string); ok {
		return str
	}

	obj, ok := body.(map[string]interface{})
	if !ok {
		return ""
	}

	// Try common error keys
	for _, key := range []string{"error", "message", "detail"} {
		if val, ok := obj[key]; ok {
			if str, ok := val.(string); ok && str != "" {
				return str
			}
		}
	}

	return ""
}

// classifyError maps HTTP status codes to typed errors.
func classifyError(statusCode int, message string, body interface{}) error {
	apiErr := APIError{
		Message:      message,
		StatusCode:   statusCode,
		ResponseBody: body,
	}

	switch statusCode {
	case http.StatusBadRequest:
		return &BadRequestError{APIError: apiErr}
	case http.StatusUnauthorized, http.StatusForbidden:
		return &AuthenticationError{APIError: apiErr}
	case http.StatusNotFound:
		return &NotFoundError{APIError: apiErr}
	case http.StatusConflict:
		return &ConflictError{APIError: apiErr}
	default:
		if statusCode >= 500 && statusCode < 600 {
			return &ServerError{APIError: apiErr}
		}
		return &apiErr
	}
}

File: ones-gfx-sdk-python/ones_gfx/transport.py
from __future__ import annotations

import json
from typing import Any
from urllib.parse import urljoin

import requests

from .auth import AuthBase
from .enums import OperationMode
from .exceptions import (
    APIError,
    AuthenticationError,
    BadRequestError,
    ConflictError,
    NotFoundError,
    ServerError,
    TransportError,
)


class Transport:
    def __init__(self, base_url: str, auth: AuthBase, verify_tls: bool | str = True, timeout: int = 30):
        if not base_url:
            raise ValueError("base_url is required.")
        self._base_url = base_url if base_url.endswith("/") else base_url + "/"
        self._auth = auth
        self._verify_tls = verify_tls
        self._timeout = timeout
        self._session = requests.Session()

    def get(self, path: str, timeout: int | None = None) -> Any:
        return self._request("GET", path, timeout=timeout)

    def post(self, path: str, json_body: dict | None = None, mode: OperationMode = OperationMode.SYNCHRONOUS, timeout: int | None = None) -> Any:
        return self._request("POST", path, json_body=json_body, mode=mode, timeout=timeout)

    def patch(self, path: str, json_body: dict | None = None, mode: OperationMode = OperationMode.SYNCHRONOUS, timeout: int | None = None) -> Any:
        return self._request("PATCH", path, json_body=json_body, mode=mode, timeout=timeout)

    def delete(self, path: str, mode: OperationMode = OperationMode.SYNCHRONOUS, timeout: int | None = None) -> Any:
        return self._request("DELETE", path, mode=mode, timeout=timeout)

    def close(self) -> None:
        self._session.close()
        self._auth.close()

    def _request(self, method: str, path: str, json_body: dict | None = None, mode: OperationMode = OperationMode.SYNCHRONOUS, timeout: int | None = None) -> Any:
        url = urljoin(self._base_url, path.lstrip("/"))
        if self._auth.needs_proactive_refresh():
            self._auth.refresh()
        response = self._send(method, url, json_body, mode, timeout)
        if response.status_code == 401:
            self._auth.refresh()
            response = self._send(method, url, json_body, mode, timeout)
        return self._handle_response(response)

    def _send(self, method: str, url: str, json_body: dict | None, mode: OperationMode, timeout: int | None) -> requests.Response:
        headers = {"Content-Type": "application/json", "Accept": "application/json"}
        self._auth.apply(headers)
        if mode is OperationMode.SYNCHRONOUS:
            headers["Prefer"] = "respond-sync"
        elif mode in (OperationMode.ASYNC_POLL, OperationMode.ASYNC_WEBHOOK):
            headers["Prefer"] = "respond-async"
        try:
            return self._session.request(
                method=method,
                url=url,
                headers=headers,
                json=json_body,
                timeout=self._timeout if timeout is None else timeout,
                verify=self._verify_tls,
            )
        except requests.exceptions.RequestException as e:
            raise TransportError(f"{method} {url} failed: {e}") from e

    def _handle_response(self, response: requests.Response) -> Any:
        body = self._parse_body(response)
        if 200 <= response.status_code < 300:
            return self._unwrap_envelope(body)
        message = self._extract_error_message(body) or f"HTTP {response.status_code}"
        exc_cls = self._classify_status(response.status_code)
        raise exc_cls(message, status_code=response.status_code, response_body=body)

    @staticmethod
    def _parse_body(response: requests.Response) -> Any:
        if not response.content:
            return None
        try:
            return response.json()
        except ValueError:
            return response.text

    @staticmethod
    def _unwrap_envelope(body: Any) -> Any:
        if not isinstance(body, dict):
            return body
        if "status" not in body or ("data" not in body and "message" not in body):
            return body
        if body.get("status") != "success":
            return body
        data = body.get("data")
        if isinstance(data, str):
            try:
                data = json.loads(data)
            except (json.JSONDecodeError, ValueError):
                pass
        if data is None:
            return body.get("message")
        return data

    @staticmethod
    def _extract_error_message(body: Any) -> str | None:
        if not isinstance(body, dict):
            return str(body) if body else None
        for key in ("error", "message", "detail"):
            value = body.get(key)
            if value:
                return str(value)
        return None

    @staticmethod
    def _classify_status(status_code: int) -> type[APIError]:
        if status_code == 400:
            return BadRequestError
        if status_code in (401, 403):
            return AuthenticationError
        if status_code == 404:
            return NotFoundError
        if status_code == 409:
            return ConflictError
        if 500 <= status_code < 600:
            return ServerError
        return APIError
